Sources/WebSocketKit/PMCE.swift

```diff
@@ -0,0 +1,753 @@
+            try decompressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+
+    }
+
+    /// websocket send calls this to compress.
+    public func compressed(_ buffer: ByteBuffer,
+                            fin: Bool = true,
+                            opCode: WebSocketOpcode = .binary) throws -> WebSocketFrame {
+        
+        guard let channel = channel else {
+            throw IOError(errnoCode: 0, reason: "PMCE: channel not configured.")
+        }
+        let notakeover = !shouldTakeOverContext()
+
+        let startSize = buffer.readableBytes
+        
+        if logging {
+            logger.trace("compressing \(startSize) bytes for \(opCode)")
+        }
+        do {
+            var mutBuffer = buffer
+
+            if !notakeover {
+                mutBuffer = unpad(buffer:buffer)
+            }
+
+            let startTime = Date()
+            
+            let compressed =
+            try mutBuffer.compressStream(with: compressorBox.value!,
+                                         flush: .sync,
+                                         allocator: channel.allocator)
+                                        //rfc says to strip padding word off compressed
+            mutBuffer = unpad(buffer: mutBuffer)
+            if logging {
+                let endTime = Date()
+                let endSize = compressed.readableBytes
+                
+                logger.trace("compressed \(startSize) to \(endSize) bytes @ \(startSize / endSize) ratio from")
+                switch extendedSocketType {
+                case .server:
+                    logger.trace(" \(serverConfig.deflateConfig.zlibConfig)")
+                case .client:
+                    logger.trace(" \(clientConfig.deflateConfig.zlibConfig)")
+                }
+                
+                logger.trace("in \(startTime.distance(to: endTime))")
+            }
+            
+            if notakeover {
+                logger.trace("resetting compressor stream")
+                try compressorBox.value?.resetStream()
+            }else {
+                logger.trace("not resetting compressor stream.")
+            }
+            
+            var frame = WebSocketFrame(
+                fin: true,
+                opcode: opCode,
+                maskKey: self.makeMaskKey(),
+                data: compressed
+            )
+            
+            frame.rsv1 = true // denotes compression
+            let slice = compressed.getSlice(at:compressed.readerIndex,
+                                                         length: compressed.readableBytes - 4)
+            frame.data = slice ?? compressed
+
+            return frame
+        }
+        catch {
+            logger.error("PMCE: send compression failed \(error)")
+        }
+        
+        return WebSocketFrame(fin:fin, rsv1: false, opcode:opCode, data: buffer)
```

Comment:
No. You're treating a compression error as "just don't compress"; this is incorrect. Let compression errors get thrown.
```suggestion
```